### DISCUSSION TITLE
conda-forge recipe

### DISCUSSION THREAD

**Discussion by MatthiasKohl** [wjakob/nanobind#80] [Ideas]
For my project, it would be great to have nanobind available through conda.
This is slightly different compared to pybind since for that only headers are needed so it's easy to manage through a build system.

OTOH, for nanobind, since the library is necessary, it's cleaner to have it properly installed (even though I'm aware that it is usually linked statically into the build).

So are you OK if I submit a conda recipe to the conda-forge feedstock?
I think the license works well for this, and I'd simply attribute the conda recipe itself to this project as well.

**Comment by wjakob** [comment]
Conda to me is an elegant way of distributing Python packages with native extension libraries and depending on precompiled binaries.

What would be the point of a conda-forge recipe for nanobind, which would only be distributed in source form?

AFAIK, these days you can mix and match and just stick a PyPI line

```
  - pip:
      - nanobind
```
into your `environment.yml`. In that case, a separate conda-forge distribution channel would just add complications.

I haven't used conda in a long time and am happy to be proven wrong. At least currently don't understand the point of the request.

**Reply by MatthiasKohl** [reply]
Yes, you can do this in `environment.yml` for creating a new environment, but you cannot put this in a conda *recipe* for the build dependencies of a conda package.

Now, to work around this, it's easy enough to have *git* as a build dependency, then clone the repo and install it during build, but I think this is less elegant since we actually have to correctly install the library.
This is different from pybind where we can simply put the headers somewhere and don't care about linking in a library at all.

**Reply by wjakob** [reply]
Ok, I understand. Feel free to make such a recipe -- it would be great if you could give me co-maintainer access to the repository.

**Reply by MatthiasKohl** [reply]
Sure I can do that.

**Comment by MatthiasKohl** [comment]
@wjakob : I've added this PR (https://github.com/conda-forge/staged-recipes/pull/20755) for the conda-forge recipe. I'll update it slightly to ensure we have python version >= 3.8.
Other than that, it's a very simple recipe and it just uses the GitHub tarball one gets from tagging versions here.

One thing I would need you to do is to comment on that PR that you are willing to be a maintainer for the recipe.
In this case, this would mainly mean bumping the version when necessary, since it's unlikely the package will break since it just uses `pip install`.
Once the PR goes through, it will automatically create a new repo on which we will both be added as maintainers and which will contain the actual recipe used for conda-forge.

**Comment by MatthiasKohl** [comment]
@wjakob : I re-tested the conda package and it's not working because the source/releases from github don't contain the robin_map sub-module. I didn't catch this before because my test environment wasn't clean, and had the sub-module already.

Would it be possible for you to just add those packages as additional tarballs to the releases here?
Unfortunately, github doesn't allow you to do this automatically, see https://github.com/dear-github/dear-github/issues/214 : this is quite annoying for a project like this one which has sub-modules and only a source release.
You basically have to do a `git clone --recurse-submodules`, then checkout the right tag, create a tarball and upload it for every new release. In the linked issue, there are github actions which would do this automatically, but I'm not sure how to set this up to run whenever a new release/tag is added.

**Reply by wjakob** [reply]
Hi @MatthiasKohl,

I would rather not mess around with the packaging process for this repository. I see two ways out:

1. The input of the conda-forge deployment process could be the PyPI wheel file instead of the GitHub .tar.gz tag file. This is basically just a ZIP file that contains all contents that would be needed to make a conda-forge build (including the tsl robin_map).

2. You could yourself download a tsl robin_map tarball corresponding to the commit ID referenced by nanobind. This project is basically static at this point, so I don't foresee this commit ID changing often. If that ever happens, a corresponding change could be made to the conda-forge build recipe.

Thanks,
Wenzel

**Reply by MatthiasKohl** [reply]
@wjakob I see, I think option 1 could work fine, but I'm not yet completely sure how the install within conda would actually work, I can try to play around with this a bit today or tomorrow.
The thing is that it's already a "built distribution", which means that you cannot simply do a "pip install" on it to copy it to the right folders etc.
If it's possible for you to add a "source distribution" / sdist to PyPI (including tsl robin_map), that would great as well, but I don't know if that's easier/same as adding this to the Github Releases.

For option 2, this would mean that there would be two files and I believe conda always needs a single source file. So then I'd have to package things into a single file anyway and host that somewhere, which defeats the point IMO.

Best
Matthias

**Reply by wjakob** [reply]
Hi Matthias,

the _nanobind_ PyPI package is a source-only wheel. So the difference between a `sdist` and `wheel`-style package is zero besides very minor stuff that actually makes `sdist` more complex (it could run potentially arbitrary code via `setup.py`, which is not allowed in a wheel)

Wenzel

**Reply by wjakob** [reply]
By the way: the downloadable tarballs on the GitHub release page are actually implicitly created by the fact that there is a tag. I cannot change them. So what you were asking previously are extra steps that would be needed after every release to also upload a second tarball including the git subrepo and somehow attach it to this auto-generated release page. This is something I would like to avoid.

**Reply by MatthiasKohl** [reply]
Hey Wenzel, I do understand those concerns. I can try to update the conda recipe and we'll see whether the reviewers there accept the changes: the issue is that almost no other package would be installed from the `wheel` since that's not how conda is set up, and they can be very idealistic/fundamentalist about these things.

I'll add a PR now and I guess we'll find out

**Reply by MatthiasKohl** [reply]
plus, there are automated tools from conda-forge which update versions, see e.g. https://github.com/conda-forge/nanobind-feedstock/pull/1

I'm not sure whether those would work with the wheel, but I guess we'll see

**Reply by wjakob** [reply]
Hi @MatthiasKohl ,

have you checked this? -> https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#source-from-multiple-sources

Apparently multiple sources are legal for a conda-forge recipe. This might be useful to fetch the tagged commit from nanobind *and* a the robin-map dependency.

Best,
Wenzel

**Reply by wjakob** [reply]
I haven't followed closely and wanted to follow up: what is the status of the conda-forge package? Does it work correctly?

**Reply by MatthiasKohl** [reply]
Hi @wjakob, sorry I completely missed your previous message.
As of now, the conda package is working since I modified the install to use git to clone and checkout the robin map repo. Git is cross platform and so this can be done easily without making the build+install process within conda complicated. 
In principle, multiple sources could be an option as well but I'm not sure if conda-forge allows this: they have strict rules about how the recipes should be written because of a lot of automation for updating versions or updating the recipes for a future format etc.
So I think the current solution is OK

**Reply by wjakob** [reply]
Great, I am happy to hear it. Thank you for setting this up.
